cjdns-core/src/announcement/serialized_ann.rs

```diff
@@ -0,0 +1,733 @@
+        pub fn parse(self) -> Result<Announcement> {
+            parser::parse(self).map_err(|e| PacketError::CannotParsePacket(e))
+        }
+
+        pub(super) fn get_hash(&self) -> Digest {
+            hash(&self.0)
+        }
+
+        pub(super) fn get_entities_bytes(&self) -> &[u8] {
+            &self.0[HEADER_SIZE..]
+        }
+
+        pub(super) fn get_header_bytes(&self) -> &[u8] {
+            &self.0[..HEADER_SIZE]
+        }
+
+        fn get_signature_bytes(&self) -> &[u8] {
+            &self.0[..SIGN_SIZE]
+        }
+
+        pub(super) fn get_pub_key_bytes(&self) -> &[u8] {
+            &self.get_signed_data()[..SIGN_KEY_SIZE]
+        }
+
+        fn get_signed_data(&self) -> &[u8] {
+            &self.0[SIGN_SIZE..]
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use crate::EncodingSchemeForm;
+        use sodiumoxide::*;
+
+        fn join(slice1: &[u8], slice2: &[u8]) -> Vec<u8> {
+            slice1.iter().chain(slice2.iter()).map(|&x| x).collect()
+        }
+
+        fn hex_to_bytes(hex_string: String) -> Vec<u8> {
+            hex::decode(hex_string).expect("invalid hex string")
+        }
+
+        #[test]
+        fn test_packet_pure_fns() {
+            init().expect("sodium init failed");
+
+            let header_data = randombytes::randombytes(HEADER_SIZE);
+            for entities_len in 0..100 {
+                let entities_data = randombytes::randombytes(entities_len);
+                let announcement_data = join(&header_data, &entities_data);
+                let packet = AnnouncementPacket::try_new(announcement_data.to_vec()).expect("invalid data len");
+
+                assert_eq!(packet.get_signed_data(), &announcement_data[SIGN_SIZE..]);
+                assert_eq!(packet.get_signature_bytes(), &header_data[..SIGN_SIZE]);
+                assert_eq!(packet.get_pub_key_bytes(), &header_data[SIGN_SIZE..][..SIGN_KEY_SIZE]);
+                assert_eq!(packet.get_header_bytes(), header_data.as_slice());
+                assert_eq!(packet.get_entities_bytes(), entities_data.as_slice());
+            }
+        }
+
+        #[test]
+        fn test_sign_check() {
+            init().expect("sodium init failed");
+
+            fn create_signed_header() -> Vec<u8> {
+                let (sodium_pk, sodium_sk) = crypto::sign::gen_keypair();
+                let header_data_to_sign = {
+                    let rest_header_data = randombytes::randombytes(HEADER_SIZE - SIGN_SIZE - SIGN_KEY_SIZE);
+                    join(sodium_pk.as_ref(), &rest_header_data)
+                };
+                let sign = crypto::sign::sign_detached(&header_data_to_sign, &sodium_sk);
+                join(sign.as_ref(), &header_data_to_sign)
+            };
+
+            for _ in 0..100 {
+                let packet = AnnouncementPacket::try_new(create_signed_header()).expect("invalid packet data len");
+                assert!(packet.check().is_ok());
+            }
```

Comment:
What's the point to execute this test 100 times? Was it intended to somehow use that `i` value to generate, for example, packets of different length?

Reply:
these tests used `randombytes` previously, so calling test 100 times seemed a good way to test checking header signature validity. I mean, it's better to test something on random data for 100 times in a code with a "for" loop rather than running a test 100 times. Yeah, we could run it just one time, but I thought that it would be more reliable to test it this way.